text/0000-pmgr.md

```diff
@@ -0,0 +1,257 @@
+    PageInfo([(PhysRange, ObjectRange); F]),
+    ObjectInfo(ObjectInfo),
+    DramReq(usize),
+    DramPages([PhysRange; G]),
+    Evict(ObjID, [ObjectRange; H])
+}
+
+enum PagerCompletion {
+    Success,
+    DramPages([PhysRange, I]),
+    Err(KernelReqError),
+}
+```
+
+The pager can directly submit information about pages and objects to the kernel before the kernel
+even asks for it, allowing for prefetching (possibly in the background, or application-driven). The
+kernel can choose to ignore these prefetching requests.
+
+## Memory Management
+
+To allow for the pager to implement some storage mechanism that does DMA transfer, the kernel allows
+for DRAM-related requests over these queues. At any time the kernel can offer up unused pages of
+DRAM for the pager to gain ownership over, and the pager can also issue a request for additional
+pages. When requesting additional pages, the kernel is allowed to refuse if low on memory. The
+kernel can also issue a request to the pager to release DRAM pages. The pager is allowed to refuse.
+For this reason, the kernel should not pass ownership of all DRAM to the pager.
+
+### Access to Physical Memory
+
+Since the pager exists in userspace, it does not have access to the DRAM pages that are given to it
+by the kernel. While in most operation, it does not need to read DRAM pages (the pager can submit
+DMA commands without access to the memory), occasionally it might want to read page contents. To do
+this, the pager can simply invent an object ID and submit PageInfo requests to the kernel to
+temporarily map pages into accessable memory.
+
+### Request Types
+
+When communicating lists of pages back and forth, both sides use an array of `PhysRange`.
+The `PhysRange` struct looks like: 
+
+```{rust}
+struct PhysRange {
+    start: u64,
+    len: u32,
+    _flags: u32,
+}
+```
+
+Thus each side can communicate sets of contiguous pages with each other. The flags field is currently unused, but when
+used should be replaced with a newtype that has the same representation as a u32.
+
+## Object Info Requests
+
+The object info struct looks like:
+
+```{rust}
+struct ObjectInfo {
+    lifetime: LifetimeType,
+    backing: BackingType,
+}
+```
+
+This struct will be expanded over-time.
+
+## Eviction
+
+To support eviction, the kernel can inform the pager that a page must by written back to stable
+storage. The pager must honor the kernel's command and must assume the page is dirty. However, this
+is meant to be used as a last resort. The kernel can also submit a *request* that the pager free up
+some number of pages. If a page returned to the kernel is dirty, the kernel will submit an
+eviction notice for that page to inform the pager that it needs to write-back that page.
+
+Note that there is a distinction between "Evict" and "Sync". The former alone requires that data be stored but not
+overwrite object data that may have been modified in-memory. This is required to support transactional semantics well.
```

Comment:
Nope! Evict never writes back to stable object memory, it must write to swap-like storage.

This is because I want these operations to be explicit, they are between the kernel and the pager, after all. The functionality you're describing is recoverable by combining Sync and Evict.

Reply:
(to clarify: this is because if we need to evict dirty pages, but they are dirty because of a transaction that has not yet committed, we cannot just write them back to stable storage or we may break the transactional semantics)

Reply:
That makes sense. I know mmap is a bad interface for databases, but I thought I should ask anyways. Maybe add that clarification in.

Reply:
> mmap is a bad interface ~for databases~

FTFY